Better Transaction Payment Errors

mzxyz opened the pull request:
Better transaction payment errors:

- Catch the specific DispatchError from `cennzx-spot / buy_fee_asset` and map the error message to code number
- Check whether `fee_exchange_asset_id` exist at the top of `buy_fee_asset` fn
- Update test cases in `transaction_payment/lib.rs` and `executor/src/lib.rs`

mzxyz committed:
initial config for tx payment error
[ci skip]

mzxyz committed:
transfer buy_fee_asset error to customised code
[ci skip]

mzxyz committed:
update test with specific error code
[ci skip]

mzxyz committed:
update test cases in executor

hoani commented on crml/cennzx-spot/src/impls.rs:
```diff
@@ -62,13 +62,20 @@ impl<T: Trait> BuyFeeAsset for Module<T> {
 	) -> Result<Self::Balance, DispatchError> {
+		// check whether exchange asset id exist
+		let fee_exchange_asset_id = exchange_op.asset_id();
+		ensure!(
```
Nice addition

hoani commented on crml/transaction-payment/src/constants.rs:
```diff
@@ -0,0 +1,53 @@
+	pub const INSUFFICIENT_SELLER_CORE_ASSET_BALANCE: u8 = 198;
+	pub const BUY_AMOUNT_NOT_POSITIVE: u8 = 199;
+	pub const INVALID_ASSET_ID: u8 = 200;
+	pub const UNKNOW_BUY_FEE_ASSET: u8 = 201;
```
`UNKNOW` -> `UNKNOWN` 😉

hoani commented on executor/src/lib.rs:
```diff
@@ -900,7 +906,10 @@ mod tests {
-		assert_eq!(r, Err(InvalidTransaction::Payment.into()));
+		assert_eq!(
+			r,
+			Err(TransactionValidityError::Invalid(InvalidTransaction::Custom(198)))
```
Is it possible to import the consts here so that:
`198` -> `INSUFFICIENT_SELLER_CORE_ASSET_BALANCE`?

mzxyz replied:
Good point

hoani reviewed (approved):
Overall LGTM, a couple of small typos to consider fixing 👍

jordy25519 commented on crml/transaction-payment/src/constants.rs:
```diff
@@ -0,0 +1,53 @@
+
+	pub fn buy_fee_asset_error_to_code(error: u8) -> u8 {
+		match error {
+			4 => INSUFFICIENT_BUYER_TRADE_ASSET_BALANCE,
```
Is it possible to import CENNZ-X error type here and match over that instead of raw number?

mzxyz replied:
Yes, using `message` will be better.

This is the struct for DispatchError:
```
pub enum DispatchError {
	/// Some error occurred.
	Other(#[codec(skip)] &'static str),
	/// Failed to lookup some data.
	CannotLookup,
	/// A bad origin.
	BadOrigin,
	/// A custom error in a module
	Module {
		/// Module index, matching the metadata module index
		index: u8,
		/// Module specific error value
		error: u8,
		/// Optional error message.
		#[codec(skip)]
		message: Option<&'static str>,
	},
}
```

So we can use `message` instead of `error` number for matching.
It will be like this:

```
pub fn buy_fee_asset_error_to_code(message: &'static str) -> u8 {
		match message {
			“InsufficientBuyerTradeAssetBalance” => INSUFFICIENT_BUYER_TRADE_ASSET_BALANCE,
                        _ => ....
}
```

jordy25519 commented on crml/transaction-payment/src/constants.rs:
```diff
@@ -0,0 +1,53 @@
+//!
+//! Transaction Payment Customized Error Code Constants
+
+pub(crate) mod error_code {
```
Did you try using `decl_error!` for this?

mzxyz replied:
The reason why I am not consider using `decl_error!` is `InvalidTransaction::Custom` only accept `u8` and also doughnut has hold some fields https://github.com/plugblockchain/plug-blockchain/blob/d2319b9c98c209be11544e5b05b4797e670c5be6/prml/doughnut/src/constants.rs#L19

so if we use `decl_error!` it will be hard to identify the `code` as for specific module the `error` code always starts from 1..n

jordy25519 replied:
Ah yes 👍

mzxyz committed:
update from comments
* replace magic number with explicit error message
* using error msg str to replace error order number

hoani reviewed (approved):
Nice changes

jordy25519 commented on crml/transaction-payment/src/lib.rs:
```diff
@@ -277,7 +288,9 @@ where
 				Ok(imbalance) => imbalance,
-				Err(_) => return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment)),
+				Err(_) => {
+					return Err(InvalidTransaction::Custom(error_code::INSUFFICIENT_SELLER_CORE_ASSET_BALANCE).into())
```
This error is abit confusing, it really means the user could not pay the transaction fee so would be good to have a simple, more meaningful error e.g.

"error_code::INSUFFICIENT_FEE_ASSET"

jordy25519 reviewed (commented):
One change requested then I think we're good to go!

mzxyz committed:
use meaningful error msg for withdraw failed

cowboy-bebug committed:
Update integration tests with crml-transaction-payment errors
Bumped max_payment for a test to avoid
ASSET_TO_CORE_PRICE_ABOVE_MAX_LIMIT error

cowboy-bebug committed:
Rename error to be more meaningful, and add some comments

jordy25519 commented on runtime/tests/tests.rs:
```diff
@@ -687,7 +696,7 @@ fn contract_call_fails_when_exchange_liquidity_is_low() {
 			initialize_block();
 			let r = Executive::apply_extrinsic(xt);
-			assert_eq!(r, Err(InvalidTransaction::Payment.into()));
+			assert_eq!(r, Err(InvalidTransaction::Custom(UNKNOWN_BUY_FEE_ASSET).into()));
```
The purpose of this PR is to make errors known so I didn't expect this 'unknown' kind to actually be used 🤔 
Should be an 'insufficient exchange liquidity for fee payment'

cowboy-bebug replied:
Yep I'm working on this right now lol thanks for the comment

jordy25519 commented on runtime/tests/tests.rs:
```diff
@@ -717,6 +726,6 @@ fn contract_call_fails_when_cpay_is_used_for_fee_exchange() {
 			});
 			let r = Executive::apply_extrinsic(xt);
-			assert_eq!(r, Err(InvalidTransaction::Payment.into()));
+			assert_eq!(r, Err(InvalidTransaction::Custom(UNKNOWN_BUY_FEE_ASSET).into()));
```
This one is a rough edge:
 - 'CannotTradeFeeAssetForItSelf' seems wordy
- Could always check this case in 'buy fee asset' and return early knowing it would result in a CPAY payment anyway but more work 🤷‍♂

cowboy-bebug committed:
Add more errors to capture in integration tests

jordy25519 reviewed (approved):
Nice work all

aliXsed commented on crml/transaction-payment/src/constants.rs:
```diff
@@ -0,0 +1,63 @@
+
+pub mod error_code {
+	// Core to asset exchange errors
+	pub const CORE_TO_ASSET_PRICE_NOT_ABOVE_ZERO: u8 = 190;
```
I know it's approved and merged. However, I believe it's good to still share our suggestions. Here, I think enum is better suited for encompassing all the possible error cases.

cowboy-bebug replied:
Yes I agree with you on this. But the reason why `u8` was used was because we have to propagate `InvalidTransaction::Custom(u8)` error type

aliXsed commented on crml/transaction-payment/src/constants.rs:
```diff
@@ -0,0 +1,63 @@
+
+	// Matches and converts crml-cennzx-spot module errors, such that
+	// they are propagated in crml-transaction-payment module
+	pub fn buy_fee_asset_error_msg_to_code(message: &'static str) -> u8 {
```
If you had the enum (Referring to my previous comment), then you may just need to implement ToString or Display for your enum. Then you should always pass the enum entry between functions until you want to log or print it in which case you would convert them to strings. So I believe converting strings to enum entries (or constants) which is not a light function (in terms of performance) could be avoidable.

cowboy-bebug replied:
This function converts string message in `DispatchError` into the appropriate `InvalidTransaction::Custom(u8)` error type

aliXsed commented on crml/transaction-payment/src/mock.rs:
```diff
@@ -142,10 +142,20 @@ impl BuyFeeAsset for Module<Runtime> {
+				return Err(DispatchError::Module {
+					index: 1,
+					error: 15,
+					message: Some("CoreToAssetPriceAboveMaxLimit"),
```
You shouldn't need to send a string like this. You should be able to return an enum entry/a defined const. Using hard-coded string literals is not a good practice.

cowboy-bebug replied:
For this we wanted to return the error explicitly for testing purposes I guess lol
